### PR TITLE
feat(stages): implement stages view

### PR DESCRIPTION
**Pull request by davidvader**
Implements viewing a pipeline that contains staged steps.
Changes can be summarized with
- Small refactor to Build.View to reduce function parameter noisiness
- When steps contain stages, group them
- Render a label for each stage, except for reserved stages like init/clone

### CODE REVIEW

**Comment by davidvader** [comment]
Screen Shot 2020-10-09 at 3 08 34 PM
Screen Shot 2020-10-09 at 3 09 13 PM

File: cypress/integration/steps.spec.js [hunk 1]
```diff
@@ -297,20 +297,12 @@ context('Steps', () => {
       cy.get('@cloneStep').contains('problem starting container');
     });
 
-    it("first step should not have 'last' styles", () => {
```
Comment (davidvader) [inline cypress/integration/steps.spec.js:300]: removing cypress tests to accommodate removing `-last` class

File: src/elm/Pages/Build/Model.elm [hunk 2]
```diff
@@ -40,6 +40,16 @@ type alias PartialModel a =
     }
 
 
+{-| BuildModel : model to contain build information that is crucial for rendering a pipeline
```
Comment (davidvader) [inline src/elm/Pages/Build/Model.elm:43]: adding an argument model so we dont have to pass 12 arguments to every single Build.View function

File: src/elm/Pages/Build/View.elm [hunk 3]
```diff
@@ -121,16 +123,16 @@ viewBuild { time, build, steps, logs, followingStep, shift } org repo =
                     )
 
         buildSteps =
-            case steps of
+            case model.steps of
                 RemoteData.Success steps_ ->
-                    viewSteps time org repo buildNumber steps_ logs followingStep shift
+                    viewPipeline model <| BuildModel org repo buildNumber steps_
```
Comment (davidvader) [inline src/elm/Pages/Build/View.elm:128]: if we construct and pass in a BuildModel here, we dont need to pass along 12 arguments for each function that is deeper.

File: src/elm/Pages/Build/View.elm [hunk 4]
```diff
@@ -222,36 +224,42 @@ viewPreview now org repo build =
         ]
 
 
-{-| viewSteps : sorts and renders build steps
+{-| viewPipeline : takes build/steps and renders pipeline
 -}
-viewSteps : Posix -> Org -> Repo -> BuildNumber -> Steps -> Logs -> Int -> Bool -> Html Msg
-viewSteps now org repo buildNumber steps logs follow shift =
+viewPipeline : PartialModel a -> BuildModel -> Html Msg
+viewPipeline model buildModel =
```
Comment (davidvader) [inline src/elm/Pages/Build/View.elm:230]: using BuildModel and PartialModel the code is much cleaner

File: src/elm/Pages/Build/View.elm [hunk 5]
```diff
@@ -222,36 +224,42 @@ viewPreview now org repo build =
...
     div [ class "steps" ]
         [ div [ class "-items", Util.testAttribute "steps" ] <|
-            List.map
-                (\step ->
-                    viewStep now org repo buildNumber step steps logs follow shift
-                )
-            <|
-                steps
+            if hasStages buildModel.steps then
```
Comment (davidvader) [inline src/elm/Pages/Build/View.elm:233]: only render grouped stages if the pipeline requires it

**Comment by JordanSussman** [comment]
Can the stages be updated to collapse/expand the steps underneath them? I'm particularly thinking of the ease of viewing the steps you care about when you potentially have a pipeline with many stages.

**Comment by davidvader** [comment]
> Can the stages be updated to collapse/expand the steps underneath them? I'm particularly thinking of the ease of viewing the steps you care about when you potentially have a pipeline with many stages.

@JordanSussman I was worried having a collapse inside of a collapse would be jarring but let me play around with it 👍

**Comment by davidvader** [comment]
Implemented very basic collapsible approach to stages and took the mockups back to the team and unfortunately we decided the multiple expansion was a poor user experience. 

We'll be moving forward with this implementation but I've made an issue to improve the experience 👍  
https://github.com/go-vela/community/issues/130

It will just take another few days to research and implement some new options for a design proposal but with this PR #216 the code that organizes steps into stages will make future improvements MUCH easier.

**Review by wass3r — APPROVED** [review]

**Review by kneal — APPROVED** [review]
LGTM 🐬

**Pull request merged**